API/BUG: .apply will correctly infer output shape when axis=1

jreback opened the pull request:
closes #16353
closes #17348
closes #17437
closes #18573
closes #17970
closes #17892
closes #17602 
closes #15628 
closes #18775 
closes #18901
closes #18919 

This fixes apply to work correctly when the returned shape mismatches the original. It will try to set the indices if it possible. Setting to a list-like with ``axis=1`` is now disallowed (but still possible if you operate *row-wise*). We were applying this inconsitently. This is of course a discouraged practice anyhow.

Prob should add some examples / update the doc-string a bit.

jreback committed:
API/BUG: .apply will correctly infer output shape when axis=1

closes #16353
closes #17348
closes #17437
closes #18573
closes #17970
closes #17892
closes #17602
closes #18775
closes #18901
closes #18919

jreback commented:
cc @TomAugspurger @jorisvandenbossche @chris-b1 
cc @jschendel

jreback commented:
@TomAugspurger @jorisvandenbossche if you have a chance

chris-b1 commented:
Unfortunately I'm guessing this will break _someone's_ code in an subtle way, but current behavior is obviously bad, so seems like a net positive.

I've never used it so not sure I fully understand the itention, but this does seem to break the `reduce` keyword, docstring says "If reduce is True a Series will always be returned, and if False a  DataFrame will always be returned."

```python
In [23]: pd.__version__
Out[23]: '0.22.0.dev0+310.gf6f0371'

In [24]: df.apply(lambda x: (1, 2, 3), axis=1, reduce=False)
Out[24]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

In [25]: df.apply(lambda x: (1, 2, 3), axis=1, reduce=True)
Out[25]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3


In [92]: pd.__version__
Out[92]: '0.21.0'

In [93]: df.apply(lambda x: (1, 2, 3), axis=1, reduce=False)
Out[93]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

In [94]: df.apply(lambda x: (1, 2, 3), axis=1, reduce=True)
Out[94]: 
0    (1, 2, 3)
1    (1, 2, 3)
2    (1, 2, 3)
3    (1, 2, 3)
4    (1, 2, 3)
5    (1, 2, 3)
dtype: object
```

jorisvandenbossche commented:
@jreback I will try to review this tomorrow

jreback commented:
i think i can simply deprecate reduce 

let me see

jreback commented:
``reduce=`` is deprecated, but I think we should actually figure out the use case and fix it. IIRC its if the passed apply function returns an empty, what do you return (a Series or a DataFrame). but need to sort thru some issues. maybe I will remove this and we can try later. (its the last commit and orthogonal).

jschendel commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -157,6 +157,80 @@ If installed, we now require:
+.. ipython:: python
+
+   df = pd.DataFrame([[1,2], [1,2]], columns=['a','b'])
+
```
Would be nice to display `df` here to show what it is prior to the `apply`, similar to what was done in the example above.

jschendel commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -157,6 +157,80 @@ If installed, we now require:
+    df.apply(lambda x: [1, 2, 3], axis=1)
+    df.apply(lambda x: [1, 2], axis=1)
+
+The returned input will also *not* return a Series with the list-wrapper as previously.
```
Could use code formatting here: \`\`Series\`\`

jschendel commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -157,6 +157,80 @@ If installed, we now require:
+   dtype: object
+
+
+New Behavior. The behaviour is consistent.
```
The second "behaviour" is using British spelling, but the first isn't.

jschendel commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -157,6 +157,80 @@ If installed, we now require:
+   dtype: object
+
+
+New Behaviour
```
Nitpick: British spelling

jschendel commented on pandas/core/frame.py:
```diff
@@ -4818,6 +4818,8 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
             reduce is True a Series will always be returned, and if False a
             DataFrame will always be returned.
+
+            .. deprecated:: 0.22.0
```
Should this be mentioned in the whatsnew?

jreback commented:
I reverted the deprecation off ``reduce``, I can deprecate but its complicated to actually replace it.

jreback commented:
thanks for comments @jschendel

jorisvandenbossche commented:
I didn't look at any code yet, but some general comments from experimenting with the branch:

* I am big +1 on making those things more consistent, but I think we should provide an alternative to achieve the original (in the cases where it changes), i.e. have a way to specify you want no inference on the result, just put the result as a scalar in a single cell. 
  We should also add documentation that explains the rules here to make this more predictable.

*   Column names of the result: using the example of the whatsnew, there is still an inconsistency in the resulting column names:

    ```
    In [1]: df = pd.DataFrame(np.random.randn(4, 3), columns=['A', 'B', 'C'])

    In [2]: df.apply(lambda x: [1, 2, 3], axis=1)
    Out[2]: 
       A  B  C
    0  1  2  3
    1  1  2  3
    2  1  2  3
    3  1  2  3

    In [3]: df.apply(lambda x: [1, 2], axis=1)
    Out[3]: 
       0  1
    0  1  2
    1  1  2
    2  1  2
    3  1  2
    ```

    So depending on the length of the list, it reuses the original column names or not. I would personally be in favor of fixing this inconsistency to not reuse the column names (although that would be an additional breaking change for the first case). 
    Related, when returning a Series it no longer uses the explicit given names of the result, if the length of the result matches that of the original frame:

    ```
    In [10]: df.apply(lambda x: pd.Series([1, 2], index=['test', 'other']), 1)
    Out[10]: 
       test  other
    0     1      2
    1     1      2
    2     1      2
    3     1      2

    In [11]: df.apply(lambda x: pd.Series([1, 2, 3], index=['test', 'other', 'cols']), 1)
    Out[11]: 
       A  B  C
    0  1  2  3
    1  1  2  3
    2  1  2  3
    3  1  2  3
    ```

    So this last example ([11]) is a regression, as before it correctly returned the specified new names.

*   `raw=True` seems to change some of the inference, while, based on the explanation in the docstring, it should only change the values passed to the function and not the inference afterwards on the result of the function.

    ```
    In [24]: df.apply(lambda x: (1, 2), axis=1)
    Out[24]: 
       0  1
    0  1  2
    1  1  2
    2  1  2
    3  1  2

    In [25]: df.apply(lambda x: (1, 2), axis=1, raw=True)
    Out[25]: 
    0    (1, 2)
    1    (1, 2)
    2    (1, 2)
    3    (1, 2)
    dtype: object
    ```

* Regarding the example of @chris-b1 above (https://github.com/pandas-dev/pandas/pull/18577#issuecomment-349994027), so for tuples this change is breaking the behaviour always (regardless of the length of the tuple). Before, tuples were always regarded as 'scalars', unless you specified `reduce=False`. 
  I am not fully sure I like this change, and as said above I think we need an alternative if keeping to this change.

*   An illustration of where this "now always try to unpack iterable things like tuples into columns" would cause a problem in a real world use case:

    This branch:

    ```
    In [53]: from shapely.geometry import MultiLineString

    In [54]: def f(row):
        ...:     return MultiLineString([[(1,2), (3,4)], [(3,4), (4,2)]])

    In [55]: df.apply(f, axis=1)
    Out[55]: 
                           0                      1
    0  LINESTRING (1 2, 3 4)  LINESTRING (3 4, 4 2)
    1  LINESTRING (1 2, 3 4)  LINESTRING (3 4, 4 2)
    2  LINESTRING (1 2, 3 4)  LINESTRING (3 4, 4 2)
    3  LINESTRING (1 2, 3 4)  LINESTRING (3 4, 4 2)
    ```

    with 0.21.1 it preserves the returned objects in a single series (the desired result):

    ```
    In [94]: df.apply(f, axis=1)
    Out[94]: 
    0    (LINESTRING (1 2, 3 4), LINESTRING (3 4, 4 2))
    1    (LINESTRING (1 2, 3 4), LINESTRING (3 4, 4 2))
    2    (LINESTRING (1 2, 3 4), LINESTRING (3 4, 4 2))
    3    (LINESTRING (1 2, 3 4), LINESTRING (3 4, 4 2))
    4    (LINESTRING (1 2, 3 4), LINESTRING (3 4, 4 2))
    5    (LINESTRING (1 2, 3 4), LINESTRING (3 4, 4 2))
    dtype: object
    ```

* I also noticed that the behaviour when returning arrays changed (to be consistent with lists, so probably the change is good), but just noticing it to make sure it is tested (previously it raised an error when the shape didn't match (eg `lambda x: np.array([1, 2])` with the example above), now it behaves like a list).

jorisvandenbossche commented:
Contemplating my above comments, I think I would be in favor of (1) "no inference" for general iterable things (objects, tuples, ..), and (2) a clear way to enable / disable inference. 

The switch to enable/disable inference could have a default depending on the result type, eg by default infer (so not putting the result as a scalar in a single cell) for Series objects and maybe arrays, and by default no inference for tuples, dicts, generic objects, and probably lists as well. When doing this, I think the main breakage is the case of a "list with length equal to number of columns of original frame", but in that case we could even catch this case and raise a FutureWarning.

Note I didn't look at all examples from the issues yet, so I might be missing some cases.

jorisvandenbossche commented:
> (and can remove the reduce= & raw= kwargs as well

I think `raw` is still orthogonal (it is to determine which type if passed to the function, not related to what is returned by the function). 
But `broadcast` *could* maybe be integrated.

jorisvandenbossche commented:
What would a `output='reduce'` do? (compared to `output='scalar'` ?)

jreback commented:
yeah going to see what I can do with all of these parameters, calling it ``result_type`` for now.

jreback commented on pandas/io/formats/style.py:
```diff
@@ -508,7 +508,8 @@ def _apply(self, func, axis=0, subset=None, **kwargs):
         data = self.data.loc[subset]
         if axis is not None:
-            result = data.apply(func, axis=axis, **kwargs)
+            result = data.apply(func, axis=axis,
```
@TomAugspurger I had to do this work-around here to push a list-like result into a frame (which we are breaking generally in this patch, but here style is relying on this, so kind of ok with this behavior, I have to add some more tests); this is from ``pandas/tests/io/formats/test_style.test_apply_axis``

. e.g.

```
In [5]:         df = pd.DataFrame({'A': [0, 0], 'B': [1, 1]})
   ...:         f = lambda x: ['val: {max}'.format(max=x.max()) for v in x]
   ...: 

In [6]: df
Out[6]: 
   A  B
0  0  1
1  0  1

In [7]: df.apply(f, axis=1)
Out[7]: 
0    [val: 1, val: 1]
1    [val: 1, val: 1]
dtype: object

In [8]: df.apply(f, axis=1, result_type='infer')
Out[8]: 
        A       B
0  val: 1  val: 1
1  val: 1  val: 1

```

I am not entirely sure why you need this this way.

jreback commented:
cc @TomAugspurger

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -4847,6 +4847,10 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
             reduce is True a Series will always be returned, and if False a
             DataFrame will always be returned.
+        result_type : str {'infer'}, optional
+            * infer : when axis=1 list-like results will be turned into columns
```
What are the valid options here? None and 'infer', or can use provide a type to ensure a specific type is returned?

Also could you clarify: is this the result_type per row, or for the entire thing? I assume per row.

TomAugspurger reviewed (approved):
Changes look good, just one question / clarification about the result_type keyword.

jreback commented:
ok this is cleaned up a bit more. I think we do need to try to rationalize all of the arguments, maybe deprecating some. can do that after.

jreback commented:
@jorisvandenbossche if any comments, merging soon.

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -4848,6 +4848,13 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
+            These only act when axis=1 {columns}
+            * infer : list-like results will be turned into columns
+            * None : list-like results will be returned as a list
+                     in a single column
```
I think we should explain in more detail what the different result types do (eg the fact that Series output is also inferred in the default case is not mentioned)

jorisvandenbossche reviewed (commented):
I like this more!

Some more comments (rerunning my comments above with the updated branch):

* With `result_type='infer'` it depends on the length of the returned list whether it is converted to frame or not:

    ```
    In [4]: df.apply(lambda x: [1, 2, 3], axis=1, result_type='infer')
    Out[4]: 
       A  B  C
    0  1  2  3
    1  1  2  3
    2  1  2  3
    3  1  2  3
    4  1  2  3
    5  1  2  3

    In [5]: df.apply(lambda x: [1, 2], axis=1, result_type='infer')
    Out[5]: 
    0    [1, 2]
    1    [1, 2]
    2    [1, 2]
    3    [1, 2]
    4    [1, 2]
    5    [1, 2]
    dtype: object
    ```
    
  I think it would be nice to also have the option to force the above to be a frame in both cases (like you would do an additional `.apply(pd.Series)`, but I find this always very ugly when I need that).
  
* The example with returning a Series with different names is still inconsistent:


    ```
    In [8]: df.apply(lambda x: pd.Series([1, 2, 3], index=['test', 'other', 'cols']), axis=1)
    Out[8]: 
       A  B  C
    0  1  2  3
    1  1  2  3
    2  1  2  3
    3  1  2  3
    4  1  2  3
    5  1  2  3

    In [9]: df.apply(lambda x: pd.Series([1, 2], index=['test', 'other']), axis=1)
    Out[9]: 
       test  other
    0     1      2
    1     1      2
    2     1      2
    3     1      2
    4     1      2
    5     1      2

    ```
    
  keeping the original names in the one case and using the returned names in the other (`result_type` has no influence on this). I think this should be consistent, and by default taking the names of the returned series seems more logical to me.
  
* We need to think about the interplay with the `reduce` keyword. Using Chris' example (https://github.com/pandas-dev/pandas/pull/18577#issuecomment-349994027), `reduce=True/False` determines the output type for empty frames. Previously those types matched the type of non empty ones as well (see that linked example, series / frame depending on reduce=False/True). Now the return type between empty and non-empty no longer matches (as the non-empty case always returns a series, `reduce` has no longer an influence on that):


    ```
    In [21]: df[:0].apply(lambda x: (1, 2, 3), axis=1, reduce=True)
    Out[21]: Series([], dtype: float64)

    In [22]: df[:0].apply(lambda x: (1, 2, 3), axis=1, reduce=False)
    Out[22]: 
    Empty DataFrame
    Columns: [A, B, C]
    Index: []
    ```
    
* How arrays are handled is still a bit different from lists. In this case, `result_type='infer'` also returns a frame for non-matching length (for list the below example it is twice a series as result):


    ```
    In [34]: df.apply(lambda x: np.array([1, 2]), axis=1)
    Out[34]: 
    0    [1, 2]
    1    [1, 2]
    2    [1, 2]
    3    [1, 2]
    4    [1, 2]
    5    [1, 2]
    dtype: object

    In [35]: df.apply(lambda x: np.array([1, 2]), axis=1, result_type='infer')
    Out[35]: 
       0  1
    0  1  2
    1  1  2
    2  1  2
    3  1  2
    4  1  2
    5  1  2
    ```
    
  I am not saying it needs to be necessarily the same as a list, but if we keep it, we should clarify this (test (might already be ok), document). Athough it will probably be easier to explain to keep it consistent with lists.

jreback commented:
start frame
```
In [1]: df = DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])

In [2]: df
Out[2]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3
```

1) above is fixed
```
In [3]: print(df.apply(lambda x: [1, 2, 3], axis=1, result_type='infer'))
   ...: print(df.apply(lambda x: [1, 2], axis=1, result_type='infer'))
   ...: 
   0  1  2
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3
   0  1
0  1  2
1  1  2
2  1  2
3  1  2
4  1  2
5  1  2
```

2) is fixed
```In [4]: print(df.apply(lambda x: pd.Series([1, 2, 3], index=['test', 'other', 'cols']), axis=1))
   ...: print(df.apply(lambda x: pd.Series([1, 2], index=['test', 'other']), axis=1))
   ...: 
   test  other  cols
0     1      2     3
1     1      2     3
2     1      2     3
3     1      2     3
4     1      2     3
5     1      2     3
   test  other
0     1      2
1     1      2
2     1      2
3     1      2
4     1      2
5     1      2
```

3) not handling anything different with reduce ATM. I agree we need to fix this, maybe just make it a kw of result_type. Not really sure of this usecase anyhow.

4) I added a test for this (though it wasn't different for boxing with array/list).

5) added example docs

jreback commented:
cc @jorisvandenbossche

jreback commented:
ping @jorisvandenbossche

jorisvandenbossche commented:
Just to note this is on my to do list to look at this again, will try to do it the latest this weekend

jorisvandenbossche commented:
> not handling anything different with reduce ATM. I agree we need to fix this, maybe just make it a kw of result_type. Not really sure of this usecase anyhow.

Maybe you didn't change any code related to `reduce`, but as far as I can see, the consequence of the changes is that `reduce` does no longer work in certain cases. See the example of Chris above (https://github.com/pandas-dev/pandas/pull/18577#issuecomment-349994027). 
I think I am OK with this case having breaking behaviour, but then IMO we should at least at the same time deprecate the `reduce` keyword, so the user gets some information on why it breaks and how he can replace the `reduce` keyword.

jreback commented:
> Maybe you didn't change any code related to reduce, but as far as I can see, the consequence of the changes is that reduce does no longer work in certain cases. See the example of Chris above (#18577 (comment)).

and we can do this as a followup. I didn't change reduce behavior at all. That is an existing issue. Of course it is true that reduce is not well tested either.

I am more looking to how to consolidate the API though. e.g. say we deprecate ``reduce``, how can we fold this into ``result_type``?

jorisvandenbossche commented on pandas/tests/frame/test_apply.py:
```diff
@@ -482,6 +486,200 @@ def test_apply_non_numpy_dtype(self):
+        # gh-17437
+        # if row shape is changing, infer it
+        df = pd.DataFrame(np.random.rand(10, 2))
+        result = df.apply(np.fft.fft, axis=0)
```
There is some inconsistency here in how we handle this compared to `axis=1` (but that is certainly something that can be discussed in a follow-up issue)

jreback replied:
I don't think there IS an inconsistency. We cannot set the column names of the result frame, UNLESS a Series is returned. Sure we *can*, but it would then be inconsistent depending on the length of the returned result (e.g. if it matches in columns we can, but then if it doesn't we cannot set them). Simply better to say that infer does not set column names at all.

jorisvandenbossche reviewed (commented):
Can you also add a test for `result_type='infer'` in case of dicts? (the keys become the column names with this branch)

I still have the feeling that we should think a bit more about the new keyword `result_type` we are adding. 
Eg to keep the existing behaviour in case of a list of equal length as the columns, you can actually use `df.apply(lambda x: [1,2,3], axis=1, broadcast=True` (just discovered now ..). So I think we should mention that in the docs. 
However, this is also something we could write as `result_type='broadcast'` (an additional point in favor of this is that currenlty `broadcast=True` and `result_type='infer'` are conflicting keywords (each of them would return different result), and it is not clear which of the two takes precedence).

jreback commented:
well I can certainly deprecate the ``broadcast`` keyword. I am not sure of the usecase of it. To be honest this function is way too complicated and we should remove as much as possible and just say, tough. This is an endless source of trouble.

jorisvandenbossche commented:
A good use case of broadcast=True is to keep the existing behaviour with lists (which I think should be mentioned in the whatsnew). Another one could be to broadcast a scalar result (which is kind of the same as `.transform` then, but transform fails on such case (it cannot broadcast))

jreback commented:
these are already the same (from your question above), meaning no inference
```
In [1]:     df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])
   ...:     df
   ...: 
Out[1]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

In [2]: df.apply(lambda x: [1, 2, 3], axis=1)
   ...: 
Out[2]: 
0    [1, 2, 3]
1    [1, 2, 3]
2    [1, 2, 3]
3    [1, 2, 3]
4    [1, 2, 3]
5    [1, 2, 3]
dtype: object

In [3]: df.apply(lambda x: [1, 2], axis=1)
Out[3]: 
0    [1, 2]
1    [1, 2]
2    [1, 2]
3    [1, 2]
4    [1, 2]
5    [1, 2]
dtype: object
```

jreback committed:
tests for dicts and result_type='infer'

jreback commented:
if any final comments. We will broad need a followup to deprecate ``reduce``.

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -259,6 +259,59 @@ Convert to an xarray DataArray
+    df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])
+    df
+
+Previous Behavior. If the returned shape happened to match the index, this would return a list-like.
```
Maybe change "match the index" to "match the *original* columns". I think the "index" you're referring to is the one inside the `.apply`, which may be confused with the original `df.index.`

"this would return a list-like" -> "this would return a DataFrame".

And maybe say, "... this would return a DataFrame. If the return shape did not match, a ``Series`` with lists was returned."

jreback replied:
done

TomAugspurger commented on pandas/core/apply.py:
```diff
@@ -19,20 +24,31 @@ def frame_apply(obj, func, axis=0, broadcast=False,
+        if broadcast is not None:
+            warnings.warn("The broadcast argument is deprecated and will "
+                          "be removed in a future version. You can specify "
+                          "result_type='broadcast' broadcast a scalar result",
```
Missing a "to" before "broadcast a scalar result"?

jreback replied:
updated

TomAugspurger reviewed (approved):
Just skimmed the docs. Looks good.

In http://pandas-docs.github.io/pandas-docs-travis/basics.html#row-or-column-wise-function-application we say

> Depending on the return type of the function passed to apply(), the result will either be of lower dimension or the same dimension.

May be worth exapnding that to a paragraph like

```
The return type of the function passed to `apply` affects the type of the ultimate output from DataFrame.apply

* If the applied function returns a ``Series``, the ultimate output is a ``DataFrame``. The columns match the index Series returned by the applied function.
* If the applied function returns any other type, the ultimate output is a ``Series``.

```

jorisvandenbossche reviewed (changes requested):
Cool, I didn't notice you add `result_type='broadcast'`! 
Can you add a test with a list of the correct length for that? (the existing tests only used functions that actually return a scalar, like np.mean)

jorisvandenbossche commented:
> if any final comments. We will broad need a followup to deprecate reduce.

I am fine with keeping that for a follow-up, *if* we plan to do that before 0.23. 

Regarding the `result_type` keyword / the possible output shape and format, I think the possible options are:

- whatever is returned from the function is regarded a 'scalar', and so result will be a Series
  -> this is the new default for eg list)
- whatever is returned from the function is tried to 'fit in' the original frame shape (preserving the column names). 
  -> So this is the new `result_type='broadcast'`
- whatever is returned we try to 'expand' if it is list-like to multiple columns (not preserving the original column names)
  -> this is the introduced `result_type='infer'`

And the default is then for list-likes option 1, and for Series option 3 (`result_type='infer'`). 

I have the feeling that it would be a nicer design to make all three options explicit option of `result_type`, so also add a keyword argument for option 1 (eg `result_type` = 'reduce' or 'scalar', ..). 

If we do that, the `result_type='infer'` also does not fully sound like the correct name for option 3. It is true that there we 'infer' the output shape, but the default is also already a kind of 'infer' (we infer which of the three options it is ..). So maybe we can go for `result_type='expand'` instead? (not fully sure it will be clearer though).

jorisvandenbossche commented:
> Cool, I didn't notice you add result_type='broadcast'!
Can you add a test with a list of the correct length for that? (the existing tests only used functions that actually return a scalar, like np.mean)

Additional case, this also seems to work for Series (broadcasting to the original shape / column names), so can add a test for that as well:

```
In [27]: df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])

In [28]: df.apply(lambda x: pd.Series([0, 1, 2], index=['other', 'col', 'names']), axis=1)
Out[28]: 
   other  col  names
0      0    1      2
1      0    1      2
2      0    1      2
3      0    1      2
4      0    1      2
5      0    1      2

In [29]: df.apply(lambda x: pd.Series([0, 1, 2], index=['other', 'col', 'names']), axis=1, result_type='broadcast')
Out[29]: 
   A  B  C
0  0  1  2
1  0  1  2
2  0  1  2
3  0  1  2
4  0  1  2
5  0  1  2
```

And maybe also a test to assert that when the list does not match the length of the column, an error is raised.

@jreback Sorry for all my comments. But this PR will really be a very nice enhancement to the clarity and consistency of apply!

The discussion of the renaming of the keyword arguments can maybe also be done in a follow-up.

jreback committed:
clean up broadcast tests

jreback committed:
rename infer to expand

jreback committed:
deprecate reduce

jreback commented:
ok deprecated ``reduce`` as well.

this prob needs slight doc followups. but is good to go.

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -793,8 +793,12 @@ The :meth:`~DataFrame.apply` method will also dispatch on a string method name.
 
-Depending on the return type of the function passed to :meth:`~DataFrame.apply`,
-the result will either be of lower dimension or the same dimension.
+The return type of the function passed to :meth:`~DataFrame.apply` affects the
```
return type -> default return type (and maybe after the two bullet points mention that it can be controlled with the `result_type` keyword (without going into details) ?)

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -793,8 +793,12 @@ The :meth:`~DataFrame.apply` method will also dispatch on a string method name.
+type of the ultimate output from DataFrame.apply
+
+* If the applied function returns a ``Series``, the ultimate output is a ``DataFrame``.
+  The columns match the index ``Series`` returned by the applied function.
```
the index Series -> the index of the Series

jreback committed:
validate result_type kwarg

jreback committed:
docs

jreback commented:
ok fixed up.

jorisvandenbossche commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -288,6 +288,73 @@ Convert to an xarray DataArray
+
+.. ipython:: python
+
+    df.apply(lambda x: [1, 2, 3], axis=1, result_type='infer')
```
small fixup needed: it's still 'infer' here (was renamed to 'expand')

jreback replied:
fixed

jreback committed:
doc

jorisvandenbossche commented:
There seems to be a "regression" compared to a previous version of the PR (or, I might incorrectly remember testing it), and it is also a change in behaviour compared to 0.22:

PR:
```
In [21]: df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])

In [26]: df.apply(lambda x: pd.Series([1, 2, 3]), axis=1, broadcast=True)
/home/joris/miniconda3/envs/dev/bin/ipython:1: FutureWarning: The broadcast argument is deprecated and will be removed in a future version. You can specify result_type='broadcast' to broadcast the result to the original dimensions
  #!/home/joris/miniconda3/envs/dev/bin/python
Out[26]: 
   0  1  2
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3
```

On pandas 0.22:
```
In [5]: df.apply(lambda x: pd.Series([1, 2, 3]), axis=1, broadcast=True)
Out[5]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3
```

And I remember a previous time testing this on the PR and thought it was working as before.

So basically the question is: if we explicitly say to broadcast, it should "preserve" the original dataframe layout, so the column names should be preserved. 
For lists that is clear, but for Series you of course have either the option to only regard the values (so the length should match) or to align. But now it is doing neither and just ignoring the `result_type='broardcast`.

jorisvandenbossche commented:
So I remembered correctly: https://github.com/pandas-dev/pandas/pull/18577#issuecomment-362410110. The behaviour of the example in that comment has changed with the latest version of the PR.

jreback committed:
broadcast should preserve like-columns

jreback commented:
your example is actually wrong, in that the result is good. it IS taking the index of the returned Series, which is [0, 1, 2]. I changed to do this

```
In [1]: df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])
   ...: 

In [2]: df.apply(lambda x: [1, 2, 3], axis=1, result_type='broadcast')
Out[2]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

In [3]: df.apply(lambda x: Series([1, 2, 3], index=list('abc')), axis=1, result_type='broadcast')
Out[3]: 
   a  b  c
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3
```

for broadcast, a list-like will get the original index, but if a Series is returned it will be the new index. I think this is consistent and useful to do.

jreback committed:
update doc-string

jorisvandenbossche commented:
> for broadcast, a list-like will get the original index, but if a Series is returned it will be the new index. I think this is consistent and useful to do.

If we decide to not broadcast a Series to the original layout, I think we should at least raise an error for that case, and not silently ignore the `result_type` keyword you passed. 
But it is a breaking change compared to 0.22 (although, I suppose this is rather a corner case .. returning series *and* using broadcast=True)

jreback committed:
test for Series with broadcast=True, raises when passed incorrect shape

jreback commented:
> If we decide to not broadcast a Series to the original layout, I think we should at least raise an error for that case, and not silently ignore the result_type keyword you passed.
But it is a breaking change compared to 0.22 (although, I suppose this is rather a corner case .. returning series and using broadcast=True)

this already raises, added a test.

jorisvandenbossche commented:
Yes, if it has an incorrect shape, then it already raises (just like lists). But my example has a correct length, only different index names.

Using your example above:

```
In [1]: df = pd.DataFrame(np.tile(np.arange(3), 6).reshape(6, -1) + 1, columns=['A', 'B', 'C'])
   ...: 

In [2]: df.apply(lambda x: [1, 2, 3], axis=1, result_type='broadcast')
Out[2]: 
   A  B  C
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3

In [3]: df.apply(lambda x: Series([1, 2, 3], index=list('abc')), axis=1, result_type='broadcast')
Out[3]: 
   a  b  c
0  1  2  3
1  1  2  3
2  1  2  3
3  1  2  3
4  1  2  3
5  1  2  3
```

I think the above is not the desired result. IMO one of the goals of  `result_type='broadcast` should be to ensure that you can be 100% sure what the result will look like (i.e. dataframe with the *original* column names). 
I don't really care about this case, so raising an error instead of preserving the original column names when the returned Series has different index names (the current behaviour on master for broadcast=True) is fine for me as well. I just think we should not silently ignore `result_type='broadcast'`. If you want the result as shown above, that is what the default of `result_type=None` already does, as well as the explicit `result_type='expand'`.

jreback commented:
you think that [3] should actually be [2], IOW we ignore if the resulting index (if any) on a broadcastable? ok I think that's fine.

jreback committed:
broadcast always returns original column names

jreback commented:
ok fixed up.

jorisvandenbossche commented:
I think that is our record of closed issues in one go :-)

Thanks @jreback !

jorisvandenbossche commented:
I opened some follow-up issues: https://github.com/pandas-dev/pandas/issues/19570, https://github.com/pandas-dev/pandas/issues/19571 
And a PR with some more doc clean-ups: https://github.com/pandas-dev/pandas/pull/19573
